File: go/save-the-cow/save_the_cow.go
// Exercise: Save the Cow, really just a word guessing game backend.
package savethecow

import (
	"errors"
	"slices"
	"strings"
)

type Game struct {
	NumGuesses int    // How many guesses we have made.
	SecretWord string // The secret word
	Guesses    []rune // Runes successfully guessed.
}

const MAX_GUESSES int = 10

// Begin a new game of save the cow.
// @param word: The secret word.
func NewGame(word string) *Game {
	return &Game{
		NumGuesses: 0,
		SecretWord: word,
		Guesses:    []rune{},
	}
}

// Guess the given rune.
// @param r: The rune to guess.
// @raises: Error if the game is over.
func (g *Game) Guess(r rune) error {
	if g.State() == "Lose" {
		return errors.New("cannot guess after the game is lost")
	}
	if g.State() == "Win" {
		return errors.New("cannot guess after the game is won")
	}
	if strings.Contains(g.SecretWord, string(r)) && !slices.Contains(g.Guesses, r) {
		g.Guesses = append(g.Guesses, r)
	} else {
		g.NumGuesses++
	}
	return nil
}

// The secret word masked with only successfully guessed letters shown
// the rest are replaced with an underscore.
// @returns: Masked version of the secret word. Same as the secret word
// if all necessary letters have been guessed.
func (g *Game) MaskedWord() string {
	var ret []rune = make([]rune, len(g.SecretWord))
	for i, rune := range g.SecretWord {
		if slices.Contains(g.Guesses, rune) {
			ret[i] = rune
		} else {
			ret[i] = '_'
		}
	}
	return string(ret)
}

// How many guesses the player has remaining.
// I think this number is off by one. I subtracted 1 for the tests.
// @returns: The number of guesses remaining (not counting this one?)
func (g *Game) RemainingGuesses() int {
	if MAX_GUESSES > g.NumGuesses {
		return MAX_GUESSES - g.NumGuesses - 1
	}
	return 0 // Game over
}

// Return current state of the game, "Win", "Lose", or "Ongoing"
// @returns: Game state, "Win" if you guessed the word, "Lose" if you didn't
// guess the word and ran out of guesses, or "Ongoing" if the game is not won/lost.
func (g *Game) State() string {
	if g.MaskedWord() == g.SecretWord {
		return "Win"
	}
	if MAX_GUESSES <= g.NumGuesses {
		return "Lose"
	}
	return "Ongoing"
}

File: php/save-the-cow/SaveTheCow.php
<?php
// Exercise: Save the Cow a guessing game.

declare(strict_types=1);

// Save the Cow Game Class
class SaveTheCow
{
    /**
     * In PHP 8.4 and newer you can use Asymmetric Property Visibility to enhance data encapsulation
     * @see https://www.php.net/manual/en/language.oop5.visibility.php#language.oop5.visibility-members-aviz
     */
    const MAX_FAILURES = 10;
    public private(set) string $state;
    public private(set) string $maskedWord;
    public private(set) int $remainingFailures;
    private array $guesses;
    private string $word;
    private bool $outOfGuesses;

    // Build the SaveTheCow object.
    // @param $word: The secret word to guess.
    public function __construct(string $word)
    {
        $this->guesses = array();
        $this->remainingFailures = self::MAX_FAILURES - 1;
        $this->outOfGuesses = false;
        $this->word = strtolower($word);
        $this->MaskWord();
        $this->CalculateState();
    }

    // Recalculate the game state Win/Lose/Ongoing
    private function CalculateState() {
        if ($this->outOfGuesses) {
            $this->state = "Lose";
        } else if ($this->maskedWord == $this->word) {
            $this->state = "Win";
        } else {
            $this->state = "Ongoing";
        }
    }

    // Guess a letter. If you guess something you already guessed that is counted as a failure. You should not call this function
    // when the game is not in the ongoing state.
    // @param $guess: The letter to guess. This should be a lower case letter between "a", and "z".
    public function guess($guess) {
        if ($this->state == "Lose") {
            throw new Exception("cannot guess after the game is lost");
        } else if ($this->state == "Win") {
            throw new Exception("cannot guess after the game is won");
        }
        if (in_array($guess, $this->guesses)) {
            $this->remainingFailures--;
        } else {
            if (!str_contains($this->word, $guess)) {
                $this->remainingFailures--;
            }
            $this->guesses[] = $guess;
        }
        if ($this->remainingFailures < 0) {
            $this->remainingFailures = 0;
            $this->outOfGuesses = true;
        }
        $this->MaskWord();
        $this->CalculateState();
    }

    // Recalculate the masked word based on what letters have been give.
    // Alphabet characters (lower case) a-z will be themself if guessed and underscore (_) if not guessed yet.
    // Non alphabetic characters pass through as if guessed.
    private function MaskWord() {
        $result = "";
        foreach (str_split($this->word) as $letter) {
            if ($letter >= "a" && $letter <= "z") {
                if (in_array($letter, $this->guesses)) {
                    $result .= $letter;
                } else {
                    $result .= "_";
                }
            } else {
                $result .= $letter;
            }
        }
        $this->maskedWord = $result;
    }
}
